### DISCUSSION TITLE
Ingest variable measurements to InfluxDB

### DISCUSSION THREAD

**Discussion by 4ley** [emqx/emqx#13918] [Q&A]
I have devices which are sending mqtt on topic `devices/{deviceId}/measurement`. The schema is like this

```
{
  "type": "object",
  "properties": {
    "timestamp":  { "type": "number" },
    "temperature": { "type": "number" },
    "humidity": { "type": "number" },
  },
  "required": ["timestamp", "temperature"],
  "additionalProperties": false
}
```

The rule is this

```
SELECT
  schema_decode('measurement', payload) as payload,
  nth(2, split(topic, '/')) as deviceId
FROM "devices/+/measurement"
```

This translates via the InfluxDB action sink to the following InfluxDB line protocol:
```
measurement,id=${deviceId} temperature=${payload.temperature},humidity=${payload.humidity} ${payload.timestamp}
```

I have two questions here:

1. Since `humidity` is optional, how do I omit `humidity=${payload.humidity}` in the line protocol if the value is not in the payload?
2. In some cases it may happen that older firmware send data that I want to discard. E.g. a device sends `capacity` in the payload. The schema while having `additionalProperties` will not allow `capacity` and will reject the whole message. I would like to have only `capacity` to be discard but the other properties still be forwarded to the sink.

**Comment by zmstone** [comment]
For Q1, try this (I did not test it though):

```
SELECT
  schema_decode('measurement', payload) as payload,
CASE
  WHEN coalesce(paylaod.humidity,-1) >= 0 THEN ',humidity=' + payload.humidity
  ELSE ''
END as humidity
```

and the template:

```
measurement,id=${deviceId} temperature=${payload.temperature}${humidity} ${payload.timestamp}
```

For Q2:
maybe write two rules, one with
`WHERE is_null_var(additionalProperties)` 
and another with
`WHERE not is_null_var(additionProperties)`

**Reply by 4ley** [reply]
Thanks for the answer! Why do you use `coalesce` instead of `is_not_null_var` like this:

```
CASE
  WHEN is_not_null_var(payload.humidity) THEN concat('humidity=', payload.humidity)
  ELSE ''
END as humidity
```

**Reply by thalesmg** [reply]
I suppose it's just a matter of style, if both work.

**Reply by zmstone** [reply]
slightly safer, because `coalesce` checks for `null`, `undefined` and empty string `''`.
